File: web/src/layer.ts
import { Tile } from './tile.js';

const TILE_SIZE = 10;
const EMPTY_TILE = new Tile();

export class Layer {
  public readonly width: number;
  public readonly height: number;
  private readonly tiles: (Tile | null)[][];
  private destroyedToTileRow = -1;

  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
    this.tiles = Array.from({ length: width }, () =>
      Array.from<Tile | null>({ length: height }).fill(null)
    );
  }

  setTile(x: number, y: number, tile: Tile): void {
    this.tiles[x]![y] = tile;
  }

  getTile(x: number, y: number): Tile {
    if (x <= this.destroyedToTileRow || y < 0 || x >= this.width || y >= this.height) {
      return EMPTY_TILE;
    }
    return this.tiles[x]?.[y] ?? EMPTY_TILE;
  }

  setDestroyedToTileRow(x: number): void {
    this.destroyedToTileRow = x;
  }

  draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, overLayers: Layer[]): void {
    const firstTileX = Math.floor(-offsetX / TILE_SIZE);
    const lastTileX = Math.floor((320 - offsetX) / TILE_SIZE);
    const firstTileY = Math.floor(-offsetY / TILE_SIZE);
    const lastTileY = Math.floor((240 - offsetY) / TILE_SIZE);

    for (let y = firstTileY; y <= lastTileY; y++) {
      for (let x = firstTileX; x <= lastTileX + 1; x++) {
        const tile = this.getTile(x, y);

        let dontDraw = false;
        for (const overLayer of overLayers) {
          const ot = overLayer.getTile(x, y);
          if (ot.isOpaque()) { dontDraw = true; break; }
        }
        if (dontDraw) continue;

        if (tile.image) {
          tile.image.draw(ctx, x * TILE_SIZE + offsetX, y * TILE_SIZE + offsetY);
        }
      }
    }
  }
}

File: wii/src/layer.cpp
#include "layer.h"
#include <cmath>

Tile Layer::emptyTile_;

Layer::Layer(int w, int h) : width(w), height(h) {
    tiles_.resize(w);
    for (int x = 0; x < w; x++) {
        tiles_[x].resize(h);
    }
}

void Layer::setTile(int x, int y, const Tile& tile) {
    if (x >= 0 && x < width && y >= 0 && y < height)
        tiles_[x][y] = tile;
}

const Tile& Layer::getTile(int x, int y) const {
    if (x <= destroyedToTileRow_ || y < 0 || x >= width || y >= height)
        return emptyTile_;
    return tiles_[x][y];
}

void Layer::setDestroyedToTileRow(int x) {
    destroyedToTileRow_ = x;
}

void Layer::draw(SDL_Renderer* renderer, int offsetX, int offsetY,
                 const std::vector<Layer*>& overLayers) const {
    int firstTileX = static_cast<int>(std::floor(static_cast<float>(-offsetX) / TILE_SIZE));
    int lastTileX = static_cast<int>(std::floor(static_cast<float>(SCREEN_WIDTH - offsetX) / TILE_SIZE));
    int firstTileY = static_cast<int>(std::floor(static_cast<float>(-offsetY) / TILE_SIZE));
    int lastTileY = static_cast<int>(std::floor(static_cast<float>(SCREEN_HEIGHT - offsetY) / TILE_SIZE));

    for (int y = firstTileY; y <= lastTileY; y++) {
        for (int x = firstTileX; x <= lastTileX + 1; x++) {
            const Tile& tile = getTile(x, y);

            bool dontDraw = false;
            for (auto* overLayer : overLayers) {
                if (overLayer->getTile(x, y).isOpaque()) { dontDraw = true; break; }
            }
            if (dontDraw) continue;

            if (tile.hasImage) {
                tile.image.draw(renderer, x * TILE_SIZE + offsetX, y * TILE_SIZE + offsetY);
            }
        }
    }
}
